lib/inventory/controller.go

```diff
@@ -0,0 +1,302 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package inventory
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/client/proto"
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/interval"
+	"github.com/gravitational/trace"
+	"github.com/siddontang/go/log"
+)
+
+// Auth is an interface representing the subset of the auth API that must be made available
+// to the controller in order for it to be able to handle control streams.
+type Auth interface {
+	UpsertNode(context.Context, types.Server) (*types.KeepAlive, error)
+	KeepAliveServer(context.Context, types.KeepAlive) error
+}
+
+type testEvent string
+
+const (
+	sshKeepAliveOk  testEvent = "ssh-keep-alive-ok"
+	sshKeepAliveErr testEvent = "ssh-keep-alive-err"
+
+	sshUpsertOk  testEvent = "ssh-upsert-ok"
+	sshUpsertErr testEvent = "ssh-upsert-err"
+
+	sshUpsertRetryOk  testEvent = "ssh-upsert-retry-ok"
+	sshUpsertRetryErr testEvent = "ssh-upsert-retry-err"
+
+	handlerStart = "handler-start"
+	handlerClose = "handler-close"
+)
+
+type controllerOpts struct {
+	serverKeepAlive time.Duration
+	testEvents      chan testEvent
+}
+
+type controllerOption func(c *controllerOpts)
+
+func withServerKeepAlive(d time.Duration) controllerOption {
+	return func(opts *controllerOpts) {
+		opts.serverKeepAlive = d
+	}
+}
+
+func withTestEventsChannel(ch chan testEvent) controllerOption {
+	return func(opts *controllerOpts) {
+		opts.testEvents = ch
+	}
+}
+
+// Controller manages the inventory control streams registered with a given auth instance. Incoming
+// messages are processed by invoking the appropriate methods on the Auth interface.
+type Controller struct {
+	store           *Store
+	auth            Auth
+	serverKeepAlive time.Duration
+	serverTTL       time.Duration
+	testEvents      chan testEvent
+	closeContext    context.Context
+	cancel          context.CancelFunc
+}
+
+func NewController(auth Auth, opts ...controllerOption) *Controller {
```

Comment:
The options are private on purpose, since they are only used in local tests.  Renamed them to `controllerTestOption` to better clarify this.  In terms of why go with this instead of a struct, adding an additional unhappy path and losing compile-time input validation for a single public argument seemed unnecessary.